src/drivers/intel/cavs/dmic.c

```diff
@@ -832,6 +805,47 @@ static inline void ipm_helper2(int source[], int *ipm, int stereo[],
 }
 #endif
 
+/* Loop number of PDM controllers in the configuration. The function
+ * checks if the controller should operate as stereo or mono left (A)
+ * or mono right (B) mode. Mono right mode is setup as channel
+ * swapped mono left.
+ */
+static int stereo_helper(int stereo[], int swap[], int di)
+{
+	int cnt;
+	int i;
+	int swap_check;
+	int ret = 0;
+
+	for (i = 0; i < DMIC_HW_CONTROLLERS; i++) {
+		cnt = 0;
+		if (dmic_prm[0]->pdm[i].enable_mic_a ||
+		    dmic_prm[1]->pdm[i].enable_mic_a)
+			cnt++;
+
+		if (dmic_prm[0]->pdm[i].enable_mic_b ||
+		    dmic_prm[1]->pdm[i].enable_mic_b)
+			cnt++;
+
+		/* Set stereo mode if both mic A anc B are enabled. */
+		cnt >>= 1;
+		stereo[i] = cnt;
+
+		/* Swap channels if only mic B is used for mono processing. */
+		swap[i] = dmic_prm[di]->pdm[i].enable_mic_b & !cnt;
```

Comment:
Seppo, something like this?
swap[i] = (dmic_prm[0]->pdm[i].enable_mic_b || dmic_prm[1]->pdm[i].enable_mic_b) && !cnt;

Reply:
Yes, it's tricky but I'm leaning towards your proposal. I need to test these mono microphone scenarios to see which one is right.

Reply:
I did this but I'm not able to test mono capture (created a mono dmic topology for it). The arecord -c 1 attempt fails in kernel driver before any request arrives to firmware. Could the new mux component convert captured mono to double-mono in a test topology to work around the apparent SOF ASoC limitation?

But since we do not have ATM mono use case requirements I think it's safe to keep this as open.